---
title: "Delta ETM"
author: 
  - Yichen Zhang^[Department of Statistics, University of British Columbia, yichen.zhang@stat.ubc.ca]
  - Yongjin P. Park^[Department of Statistics, University of British Columbia, ypp@stat.ubc.ca]
date: "`r format(Sys.time(), ' %X, %b %d, %Y')`"
bibliography: "DeltaETM.bib"
csl: template/nature.csl
fontsize: 11pt
output:
  pdf_document:
    latex_engine: xelatex
    citation_package: natbib
    fig_caption: true
tables: yes
always_allow_html: true
header-includes:
 - \usepackage{fontspec}
 - \usepackage{booktabs}
 - \usepackage{longtable}
 - \usepackage{array}
 - \usepackage{multirow}
 - \usepackage{wrapfig}
 - \usepackage{float}
 - \usepackage{colortbl}
 - \usepackage{pdflscape}
 - \usepackage{tabu}
 - \usepackage{threeparttable}
 - \usepackage{threeparttablex}
 - \usepackage{makecell}
 - \usepackage{amsmath}
 - \usepackage{amssymb}
 - \usepackage{setspace}\doublespacing
---

# Background/objectives

Single-cell RNA-seq technology has been successfully applied to profile regulatory-genomic changes in studying many human diseases mechanisms. Our capability to measure single-cell-level DNA/mRNA molecules has dramatically changed our research paradigm in genomics and translational medicine. We now have a tool to accurately identify and predict the cellular contexts of aberrant samples, which confer an essential basis for understanding disease mechanisms, overcoming one of the most critical challenges in precision medicine. A typical single-cell study implicitly assumes observed transcript levels as a static value, considering that every cell is fixed at a particular steady-state. Recently, researchers have developed a complementary method to measure gene expression dynamics (the speed of RNA splicing) by taking the difference between the spliced and unspliced counts in scRNA-seq profiles.

However, the RNA velocity approach carries several technical drawbacks, primarily due to high drop-out rates in single-cell sequencing technology. Dynamic directions between cell states are usually sensitive to the choice of computational methods and tunable parameters. More importantly, most single-cell datasets, especially those collected from patient-derived cancer samples, only span over several snapshots of full developmental, evolutionary or disease progression processes. Such a shortened period in a study design may not provide enough information for full-scale dynamic modelling based on a system of ordinary differential equations (ODEs). Discontinuity and sparsity in data collection somewhat enforce statistical inference algorithms to rely on unrealistic steady-state assumptions and interpolated data points with a high level of uncertainty.

Nevertheless, gene expression dynamics implicated by the transcript-level difference between the spliced and unspliced counts provide a valuable perspective in single-cell data analysis, making single-cell analysis even more, richer and fuller. To overcome the limitations of poor and incompleteness in single-cell RNA velocity analysis, we propose a new modelling framework, Delta-ETM, short for Dynamically-Encoded Latent Transcriptomic Analysis by Embedded Topic Modelling. Delta-ETM combines two ideas: (1) latent topic analysis that will guide unsupervised machine learning for discovering new dynamic cell states, (2) application of first-order approximation to learn robust relationships between the spliced and unspliced counts instead of estimating a full trajectory of ODE models. For a latent topic model, we view each cell as a document and each gene as a word to make model parameters directly interpretable while keeping the Bayesian model's capability to impute missing information. The simplified dynamic model also permits an intuitive interpretation of spliced-unspliced differences as multiplicative "delta" parameters in the model.

# Method

## Preliminaries

Motivating example balabala.... $N$ Cells $n$, $G$ Genes $g$, $d$ Topics $t$

-   For the spliced count, $$\mathbb{E}\left[X_{ng}^{spliced}|X_{ng}^{unspliced}, Z\right] = \sum_{t}h_{nt}\left(Z\right)f_{\theta}\left(\rho_{tg} +\delta_{tg} \right)$$

-   For the unspliced count, $$\mathbb{E}\left[X_{ng}^{unspliced}|Z\right] = \sum_{k}h_{nt}\left(Z\right)f_{\theta}\left(\rho_{tg}\right)$$

$f(\rho_{tg})k()$ Softplus:

-   $Softmax(log(1+\exp(\delta))\rho$

-   $h = softmax(0.5 z_s + 0.5 z_u)$ versus concat.

-   $softmax(\rho) softmax(\delta)$

-   torch.seed()

-   nLV: 5, $\leq 30$

-   wrapper: gradient of weight on the decoder side

-   Log h every (K \~ 100) epochs

-   

    ## seperate loglikihood every K(100) epochs

## Generative process

DeltaETM posit that observed UMI counts for cell $n$ and gene $g$, are generated by the following process

```{=tex}
\begin{align}
z_{n} & \sim \operatorname{Normal}(0, I) \\
h_{n} & \sim \operatorname{Softmax }\left(z_{n}\right) \\
\rho_{g}  &\text{\quad Gene-level transcioptional parameter }\\
\delta_{t g} &\text{\quad Gene-by-topic post-transcriptional modification }\\
x_{n g} & \sim \text { ObservationModel }\left(h_{n}, \rho_{g}, \delta_{tg}\right)
\end{align}
```
The generative process of DeltaETM uses two neural networks:

```{=tex}
\begin{align}
&w_{tg} = \rho_{g} + \delta_{tg} \\
&f_{\theta}\left(\rho, \delta\right) =  \sigma(\rho_{g} + \delta_{tg})_{t} = \frac{e^{w_{t}}}{\sum_{g=1}^{G} e^{w_{tg}}} \\
&f_{w}\left(z_{n}\right): \mathbb{R}^{d} \rightarrow \mathbb{R}^{g} \\
&f_{h}\left(z_{n}\right): \mathbb{R}^{d}  \rightarrow(0,1)^{T}\\
\end{align}
```
The latent variables, along with their description are summarized in the following table:

| Latent Variables                    | Descriptions                                                                                                                                 | Code variable (if different) |
|------------------------|------------------------|------------------------|
| $z_{n} \in \mathbb{R}^{d}$          | Low-dimensional representation capturing the state of a cell.                                                                                |                              |
| $h_{n} \in \Delta^{d-1}$            | Topic proportion of of a cell, this is a vector sums up to within a cell, which is obtained by applyinng $\operatorname{Softmax}$ to $z_{n}$ |                              |
| $\rho \in \mathbb{R}^{G}$           | Gene-level transcioptional factor shared between unspliced and unspliced counts for each gene                                                |                              |
| $\delta \in \mathbb{R}^{d\times G}$ | Post-transcriptional modification per topic for each gene                                                                                    |                              |
## Some implications

When we only have gene-level effect, for example in reconstructing unspliced count, it means all topics has the same gene loadings in the unspliced counts.

# Results 

We applied our Delta-ETM approach to comprehensive single-cell datasets on pancreatic ductal adenocarcinoma (PDAC), one of the most challenging cancer types with a poor prognosis. In the latent space, we identified tumour-specific and stage-specific topics marked by a unique set of genes differentially expressed. We also found Delta-ETM further dissected sub-topics that were clumped together in traditional clustering methods implicating novel gene modules and cell states that are dynamically controlled along with the cancer progressions. \# Conclusions
